File: framework/languages/java/samples/java/TicTacToe.SessionGateway/Server/Play/src/main/java/systems/zlink/samples/tictactoe/sessiongateway/server/play/gamespots/GameNotificationPublisher.java
package systems.zlink.samples.tictactoe.sessiongateway.server.play.gamespots;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletionStage;
import systems.zlink.framework.actors.ZLinkBoundSession;
import systems.zlink.samples.tictactoe.sessiongateway.shared.configuration.SampleNames;

public final class GameNotificationPublisher {
    private GameNotificationPublisher() {
    }

    public static String encode(TicTacToeGameSpot.JoinResult result) {
        return encode(result.encode(), result.events());
    }

    public static String encode(TicTacToeGameSpot.MoveResult result) {
        return encode(result.encode(), result.events());
    }

    public static PlayResponse decode(String value) {
        String[] lines = value.split("\\n", -1);
        List<Notification> notifications = new ArrayList<>();
        for (int index = 1; index < lines.length; index++) {
            if (lines[index].isBlank()) {
                continue;
            }
            String[] parts = lines[index].split("\\|", 3);
            notifications.add(new Notification(parts[0], parts[1], parts[2]));
        }
        return new PlayResponse(lines[0], notifications);
    }

    private static String encode(String reply, List<TicTacToeGameSpot.GameEvent> events) {
        List<String> lines = new ArrayList<>();
        lines.add(reply);
        for (TicTacToeGameSpot.GameEvent event : events) {
            lines.add(packetName(event) + "|" + event.recipientActorId() + "|" + payload(event));
        }
        return String.join("\n", lines);
    }

    private static String packetName(TicTacToeGameSpot.GameEvent event) {
        return switch (event.kind()) {
            case "OpponentJoined" -> SampleNames.OpponentJoinedPacket;
            case "GameEnded" -> SampleNames.GameEndedPacket;
            default -> SampleNames.TurnChangedPacket;
        };
    }

    private static String payload(TicTacToeGameSpot.GameEvent event) {
        return event.state().encode();
    }

    public CompletionStage<Void> publishAsync(ZLinkBoundSession session, String state) {
        return session.send(state)
            .packetName("GameStateChanged")
            .submitAsync();
    }

    public record PlayResponse(String reply, List<Notification> notifications) {
    }

    public record Notification(String packetName, String recipientActorId, String payload) {
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe.SessionGateway/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/sessiongateway/server/play/gamespots/GameNotificationPublisher.kt
package systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.play.gamespots

import kotlinx.coroutines.future.await
import systems.zlink.framework.actors.ZLinkBoundSession
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.shared.configuration.SampleNames

object GameNotificationPublisher {
    fun encode(result: TicTacToeGameSpot.JoinResult): String =
        encode(result.encode(), result.events)

    fun encode(result: TicTacToeGameSpot.MoveResult): String =
        encode(result.encode(), result.events)

    fun decode(value: String): PlayResponse {
        val lines = value.split("\n")
        val notifications = lines.drop(1)
            .filter { it.isNotBlank() }
            .map {
                val parts = it.split("|", limit = 3)
                Notification(parts[0], parts[1], parts[2])
            }
        return PlayResponse(lines[0], notifications)
    }

    private fun encode(reply: String, events: List<TicTacToeGameSpot.GameEvent>): String =
        buildList {
            add(reply)
            events.forEach { event ->
                add("${packetName(event)}|${event.recipientActorId}|${event.state.encode()}")
            }
        }.joinToString("\n")

    private fun packetName(event: TicTacToeGameSpot.GameEvent): String =
        when (event.kind) {
            "OpponentJoined" -> SampleNames.OpponentJoinedPacket
            "GameEnded" -> SampleNames.GameEndedPacket
            else -> SampleNames.TurnChangedPacket
        }

    suspend fun publish(session: ZLinkBoundSession, state: String) {
        session.send(state)
            .packetName("GameStateChanged")
            .submitAsync()
            .await()
    }

    data class PlayResponse(
        val reply: String,
        val notifications: List<Notification>,
    )

    data class Notification(
        val packetName: String,
        val recipientActorId: String,
        val payload: String,
    )
}
